# comparing two types of posts  to determine which one receive more comments on average and certain time receive more comments on average

In this project, we'll aim to We'll compare two types of posts to determine whichi one receive more comments on average and posts created at a certain time receive more comments on average,we're working on Ask HN and Show HN posts.
We'll analyze existing data about user-submitted stories to   to determine which one receive more comments and certain time receive more comments . To make our recommendation, we'll try to find out:

# Exploring Existing Data 

In [2]:
#read in data
from csv import reader
open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

# Quick exploration of the data
print(hn[:5])


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [3]:
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


# Separate posts beginning with Ask HN and Show HN

In [4]:
ask_posts = []
show_posts = []
other_posts = []
# Separate posts start with Ask HN and Show HN
for row in hn :
    title = row[1].lower()
    if title.startswith('ask hn') :
        ask_posts.append(row)
    elif title.startswith('show hn') :
        show_posts.append(row)   
    else :
        other_posts.append(row)   
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


#  determine which posts receive more comments on average

In [5]:
# Compute the average number of comments on ask posts
total_ask_comments = 0
for post in ask_posts :
    ask_num_comments = int(post[4])
    total_ask_comments +=  ask_num_comments
avg_ask_comments =  total_ask_comments / len(ask_posts)
print(avg_ask_comments)
    

14.038417431192661


In [6]:
# Compute the average number of comments on show posts
total_show_comments = 0
for post in show_posts :
    show_num_comments = int(post[4])
    total_show_comments +=  show_num_comments
avg_show_comments =  total_show_comments / len(show_posts)
print(avg_show_comments)
    

10.31669535283993


The ask posts receive more comments than show posts.Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

#  determine time are more likely to attract comment

In [7]:
import datetime as dt
result_list = []
for post in ask_posts :
    post_date = post[6]
    num_comments= int(post[4])
    result_list.append([post_date, num_comments])
    

In [8]:

counts_by_hour = {}
comments_by_hour = {}
date_format = '%m/%d/%Y %H:%M'
for row in result_list :
    date = row[0]
    comment = row[1]
    post_time = dt.datetime.strptime(date, date_format)
    post_time = post_time.strftime('%H')
    if  post_time not in counts_by_hour :
        counts_by_hour[post_time] = 1
        comments_by_hour[post_time] = comment
    else :    
        counts_by_hour[post_time] += 1
        comments_by_hour[post_time] += comment
comments_by_hour        

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

In [9]:
avg_by_hour = []
for hr in comments_by_hour :
    avg_by_hour.append([hr,comments_by_hour[hr]/counts_by_hour[hr]]) 
avg_by_hour    
    

[['16', 16.796296296296298],
 ['19', 10.8],
 ['11', 11.051724137931034],
 ['04', 7.170212765957447],
 ['01', 11.383333333333333],
 ['17', 11.46],
 ['10', 13.440677966101696],
 ['02', 23.810344827586206],
 ['03', 7.796296296296297],
 ['00', 8.127272727272727],
 ['15', 38.5948275862069],
 ['09', 5.5777777777777775],
 ['20', 21.525],
 ['06', 9.022727272727273],
 ['18', 13.20183486238532],
 ['21', 16.009174311926607],
 ['07', 7.852941176470588],
 ['08', 10.25],
 ['13', 14.741176470588234],
 ['12', 9.41095890410959],
 ['22', 6.746478873239437],
 ['05', 10.08695652173913],
 ['23', 7.985294117647059],
 ['14', 13.233644859813085]]

In [10]:
swap_avg_by_hour = []
for row in avg_by_hour :
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)    
    
    

[[16.796296296296298, '16'], [10.8, '19'], [11.051724137931034, '11'], [7.170212765957447, '04'], [11.383333333333333, '01'], [11.46, '17'], [13.440677966101696, '10'], [23.810344827586206, '02'], [7.796296296296297, '03'], [8.127272727272727, '00'], [38.5948275862069, '15'], [5.5777777777777775, '09'], [21.525, '20'], [9.022727272727273, '06'], [13.20183486238532, '18'], [16.009174311926607, '21'], [7.852941176470588, '07'], [10.25, '08'], [14.741176470588234, '13'], [9.41095890410959, '12'], [6.746478873239437, '22'], [10.08695652173913, '05'], [7.985294117647059, '23'], [13.233644859813085, '14']]


In [15]:
 sorted_swap = sorted(swap_avg_by_hour ,reverse= True)

In [20]:
print( "Top 5 Hours for Ask Posts Comments") 
for avg,hr in sorted_swap[:5] :
    print("{}: {:.2f} average comments per post"
          .format(dt.datetime.strptime(hr,'%H') .strftime('%H:%M')
                  ,avg))


Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


* The hour that receives the most comments per post on average is   15:00, with an average of 38.59 comments per post.
* according documentation 15:00 (time zone US) equivalent 9:00 pm in cairo  